crates/testing/src/block_builder/mod.rs

```diff
@@ -0,0 +1,177 @@
+where
+    <TYPES as NodeType>::InstanceState: Default,
+{
+    type Config: Default;
+
+    async fn start(
+        num_storage_nodes: usize,
+        options: Self::Config,
+        changes: HashMap<u64, BuilderChange>,
+    ) -> (Box<dyn BuilderTask<TYPES>>, Url);
+}
+
+pub trait BuilderTask<TYPES: NodeType>: Send + Sync {
+    fn start(
+        self: Box<Self>,
+        stream: Box<dyn Stream<Item = Event<TYPES>> + std::marker::Unpin + Send + 'static>,
+    );
+}
+
+/// Entry for a built block
+#[derive(Debug, Clone)]
+struct BlockEntry<TYPES: NodeType> {
+    metadata: AvailableBlockInfo<TYPES>,
+    payload: Option<AvailableBlockData<TYPES>>,
+    header_input: Option<AvailableBlockHeaderInput<TYPES>>,
+}
+
+/// Construct a tide disco app that mocks the builder API.
+///
+/// # Panics
+/// If constructing and launching the builder fails for any reason
+pub fn run_builder_source<TYPES, Source>(
+    url: Url,
+    mut change_receiver: Receiver<BuilderChange>,
+    source: Source,
+) where
+    TYPES: NodeType<Transaction = TestTransaction>,
+    <TYPES as NodeType>::InstanceState: Default,
+    Source: Clone + Send + Sync + tide_disco::method::ReadState + 'static,
+    <Source as ReadState>::State: Sync + Send + BuilderDataSource<TYPES>,
+{
+    async_spawn(async move {
+        let start_builder = |url: Url, source: Source| -> _ {
+            let builder_api = hotshot_builder_api::builder::define_api::<Source, TYPES, Version01>(
+                &Options::default(),
+            )
+            .expect("Failed to construct the builder API");
+            let mut app: App<Source, Error> = App::with_state(source);
+            app.register_module::<Error, Version01>("block_info", builder_api)
+                .expect("Failed to register the builder API");
+            async_spawn(app.serve(url, STATIC_VER_0_1))
+        };
+
+        let mut handle = Some(start_builder(url.clone(), source.clone()));
+
+        while let Ok(event) = change_receiver.recv().await {
+            match event {
+                BuilderChange::Up if handle.is_none() => {
+                    handle = Some(start_builder(url.clone(), source.clone()));
+                }
+                BuilderChange::Down => {
+                    if let Some(handle) = handle.take() {
+                        #[cfg(async_executor_impl = "tokio")]
+                        handle.abort();
+                        #[cfg(async_executor_impl = "async-std")]
+                        handle.cancel().await;
+                    }
+                }
+                _ => {}
+            }
+        }
+    });
+}
+
+/// Helper function to construct all builder data structures from a list of transactions
+async fn build_block<TYPES: NodeType>(
+    transactions: Vec<TYPES::Transaction>,
+    num_storage_nodes: usize,
+    pub_key: TYPES::BuilderSignatureKey,
+    priv_key: <TYPES::BuilderSignatureKey as BuilderSignatureKey>::BuilderPrivateKey,
+) -> (
```

Comment:
Building a block is pretty crucial, so I think that perhaps it'd be better to have this return type be something other than a tuple. Can we just whip up a struct or use the `BlockEntry` struct that we use elsewhere?

Reply:
`BlockEntry` makes sense here, I'll change it